### PR TITLE
Add Auto lacing

### PR DESCRIPTION
**Pull request by ke-yu**
### Purpose

Under the hood Dynamo uses replication guide to implement different lacing strategies - 
- Shortest lacing, which currently means no replication guide
- Longest lacing. It converts a function call `foo(x, y)` to `foo(x<1L>, y<1L>)`
- Cross product. It converts a function call `foo(x, y)` to `foo(x<1>, y<2>)`

But shortest lacing is problematic. The correct semantic of shortest lacing should be `foo(x<1>, y<1>)`. This PR updates shortest lacing, and add `Auto` lacing which lets Dynamo do default replication (no replication guide). 

untitled

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@sharadkjaiswal

### FYIs

@Racel

### CODE REVIEW

**Comment by Racel** [comment]
Thanks @ke-yu. Looks good. I am still a bit concerned that the symbol we used will give the wrong impression, and am thinking maybe just writing "Auto" will make more sense on the node (but it will be more distracting). I'm going to run it by a few customers to see what they think...More soon.

File: test/core/dsfunction/contains.dyn [hunk 1]
```diff
@@ -1,19 +1,19 @@
 <Workspace Version="1.1.1.1960" X="0" Y="0" zoom="1" Name="Home" Description="" RunType="Automatic" RunPeriod="1000" HasRunWithoutCrash="True">
   <NamespaceResolutionMap />
   <Elements>
-    <Dynamo.Graph.Nodes.ZeroTouch.DSFunction guid="928ae9e8-26a8-4f5c-87c1-170688896ab1" type="Dynamo.Graph.Nodes.ZeroTouch.DSFunction" nickname="Point.ByCoordinates" x="375" y="103" isVisible="true" isUpstreamVisible="true" lacing="Shortest" isSelectedInput="False" IsFrozen="false" isPinned="false" assembly="ProtoGeometry.dll" function="Autodesk.DesignScript.Geometry.Point.ByCoordinates@double,double">
+    <Dynamo.Graph.Nodes.ZeroTouch.DSFunction guid="928ae9e8-26a8-4f5c-87c1-170688896ab1" type="Dynamo.Graph.Nodes.ZeroTouch.DSFunction" nickname="Point.ByCoordinates" x="375" y="103" isVisible="true" isUpstreamVisible="true" lacing="Auto" isSelectedInput="False" IsFrozen="false" isPinned="false" assembly="ProtoGeometry.dll" function="Autodesk.DesignScript.Geometry.Point.ByCoordinates@double,double">
```
Comment (sharadkjaiswal) [inline test/core/dsfunction/contains.dyn]: I'll rather leave some nodes with Shortest lacing and let migration code to handle it. Do we have test cases for migration?

**Comment by Racel** [comment]
@ke-yu - Something that I thought about just now. Please ensure that when "Use Levels" is applied that the lacing automatically switches to "shortest"...I don't know if this is part of this task or the should be a separate one.

For reference, we did this in https://github.com/DynamoDS/Dynamo/pull/7009

**Comment by ke-yu** [comment]
@Racel thanks for reminding. I'll include it in this PR.

File: test/core/dsfunction/contains.dyn (hunk 1 above)
Comment (ke-yu) [inline test/core/dsfunction/contains.dyn]: This one should be automatically migrated. I'll test that. :-)

**Comment by ke-yu** [comment]
@sharadkjaiswal I've reverted all changes to test .dyn files, they will be automatically migrated during testing. :-)

**Comment by ke-yu** [comment]
@Racel we probably won't change lacing to "shortest" automatically, the difference is shown below (**`t1@-2<1> + t2` v.s. `t1@-2<1> + t2<1>`**):

untitled

**Comment by Racel** [comment]
@ke-yu: You are right. We only replicate on the items that use list@level. Ok, so, maybe we just leave everything Auto.

**Comment by Racel** [comment]
@ke-yu - After  confirming with some customers, I think the auto-lacing icon we used is giving the wrong impression. Can we make a change to the icon as soon as possible? In this sprint or next? Here is a link to the new task that I assigned to @ke-yu: https://jira.autodesk.com/browse/DYN-302

**Pull request merged**